Improvements to VCS setup

hubertp opened the pull request:
### Pull Request Description

The presence of `.git` in Enso's data directory (`.enso`), which should be under VCS as well, turned out to be really problematic. Git (and JGit) treat `.enso` as a submodule which prevents as from manual inspection of the repo. Additionally, every commit changed the submodule version, which in turn was reporting invalid/changed status always.

This change renames `.enso/.git` to `.enso/.vcs` which eliminates a lot of problems. Additionally, rather than doing `git add .` we selectively add files to the index, thus preventing most of the issues reported in the ticket.
This change also means that JGit's status needs to filter `.enso/.vcs` as those will be marked as untracked always.

### Important Notes

To manually test the status of the project's repo do
```
git --git-dir .enso/.vcs status
```

### Checklist

Please include the following checklist in your PR:

- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

hubertp committed:
Improvements to VCS setup

The presence of `.git` in Enso's data directory (`.enso`), which should
be under VCS as well, turned out to be really problematic.
Git (and JGit) treat `.enso` as a submodule which prevents as from
manual inspection of the repo. Additionally, every commit changed the
submodule version, which in turn was reporting invalid/changed status
always.

This change renames `.enso/.git` to `.enso/.vcs` which eliminates a lot
of problems. Additionally, rather than doing `git add .` we selectively
add files to the index, thus preventing most of the issues reported in
the ticket.
This change also means that JGit's status needs to filter `.enso/.vcs`
as those will be marked as untracked always.
To manually test the status of the project's repo do
```
git --git-dir .enso/.vcs status
```

4e6 commented on engine/language-server/src/main/scala/org/enso/languageserver/vcsmanager/Git.scala:
```diff
@@ -78,11 +80,18 @@ private class Git(ensoDataDirectory: Option[Path]) extends VcsApi[BlockingIO] {
-          .reset()
-          .addPath(gitDir.toString)
+      val isDataDir =
+        (x: Path) => ensoDataDirectory.map(_ == x).getOrElse(false)
```
```suggestion
        ensoDataDirectory.contains _
```

4e6 commented on engine/language-server/src/main/scala/org/enso/languageserver/vcsmanager/Git.scala:
```diff
@@ -78,11 +80,18 @@ private class Git(ensoDataDirectory: Option[Path]) extends VcsApi[BlockingIO] {
+      val filesToAdd =
+        listDirectoryFiles(root, Set(gitDir)).filterNot(isDataDir)
+      println("FilesToAdd: " + filesToAdd) // for debugging windows
+      if (!filesToAdd.isEmpty) {
```
```suggestion
      if (filesToAdd.nonEmpty) {
```

4e6 commented on engine/language-server/src/main/scala/org/enso/languageserver/vcsmanager/Git.scala:
```diff
@@ -106,19 +115,29 @@ private class Git(ensoDataDirectory: Option[Path]) extends VcsApi[BlockingIO] {
+      val status    = jgit.status.call()
+      val untracked = status.getUntracked()
+      val isDataDir = (x: String) =>
+        ensoDataDirectory.map(dataDir => dataDir.toString == x).getOrElse(false)
```
```suggestion
        ensoDataDirectory.exists(dataDir => dataDir.toString == x)
```

4e6 commented on engine/language-server/src/main/scala/org/enso/languageserver/vcsmanager/Git.scala:
```diff
@@ -106,19 +115,29 @@ private class Git(ensoDataDirectory: Option[Path]) extends VcsApi[BlockingIO] {
+      }
+      if (!filesToAdd.isEmpty) {
+        val addCmd = jgit.add()
+        filesToAdd.foreach(addCmd.addFilepattern(_))
```
```suggestion
        filesToAdd.foreach(addCmd.addFilepattern)
```

4e6 reviewed (approved):
:+1:

hubertp committed:
Ensure `/` separator in AddCommand.addFilePattern

That's what the docs say and CI picked it up

hubertp committed:
JGit StatusCmd returns results that are unix-based

Despite running on Windows, the results have a unix-based, `/`,
separator. This was causing problems when removing uninteresting
results.

hubertp committed:
Fix windows/unix separator in tests

hubertp committed:
Apply suggestions from code review

Co-authored-by: Dmitry Bushev <bushevdv@gmail.com>

hubertp committed:
Merge branch 'develop' into wip/hubert/git-add-fix-183992502
